File: algorithms/visualstudio/algorithms.csharp/DivideAndConquer/KaratsubaMultiplication.cs
using System.Numerics;

namespace algorithms.csharp.DivideAndConquer
{
    public class DividedNumber
    {
        private readonly BigInteger _highNumber;
        private readonly BigInteger _lowNumber;

        public DividedNumber(BigInteger highNumber, BigInteger lowNumber)
        {
            _highNumber = highNumber;
            _lowNumber = lowNumber;
        }

        public BigInteger HighNumber { get { return _highNumber; } }
        public BigInteger LowNumber { get { return _lowNumber; } }
    }

    public class KaratsubaMultiplication
    {
        public int DividePoint(BigInteger number)
        {
            return number.ToString().Length / 2;
        }

        public DividedNumber SplitNumber(BigInteger number, int splitAt)
        {
            var splitBase = BigInteger.Pow(MakeBigInteger(10), splitAt);
            
            var high = number / splitBase;
            var low = number % splitBase;

            return new DividedNumber(high, low);
        }

        public BigInteger Multiply(BigInteger firstNumber, BigInteger secondNumber)
        {
            var max = firstNumber > secondNumber ? firstNumber : secondNumber;

            if (max < 10)
            {
                return firstNumber * secondNumber;
            }

            return MultiplyRecursion(max, firstNumber, secondNumber);
        }

        private BigInteger MultiplyRecursion(BigInteger max, BigInteger firstNumber, BigInteger secondNumber)
        {
            var midPoint = DividePoint(max);

            var splitFirstNumber = SplitNumber(firstNumber, midPoint);
            var splitSecondNumber = SplitNumber(secondNumber, midPoint);

            var highs = Multiply(splitFirstNumber.HighNumber, splitSecondNumber.HighNumber);
            var lows = Multiply(splitFirstNumber.LowNumber, splitSecondNumber.LowNumber);

            var sumOfSplitFirstNumber = splitFirstNumber.HighNumber + splitFirstNumber.LowNumber;
            var sumOfSplitSecondNumber = splitSecondNumber.HighNumber + splitSecondNumber.LowNumber;
            var mids = Multiply(sumOfSplitFirstNumber, sumOfSplitSecondNumber);

            var firstTerm = highs*(BigInteger.Pow(10, (2*midPoint)));
            var midTerm = (mids - highs - lows)*(BigInteger.Pow(10, midPoint));
            var lastTerm = lows;

            return firstTerm + midTerm + lastTerm;
        }


        private BigInteger MakeBigInteger(int number)
        {
            return new BigInteger(number);
        }
    }
}

File: algorithms/visualstudio/algorithms.fsharp/KaratsubaMultiplication.fs
namespace algorithms.fSharp.divideAndConquer

open System.Numerics

module KaratsubaMultiplication = 
    
    let dividePoint (number: BigInteger) = 
        let numberLength = number.ToString()
                           |> Seq.length

        numberLength / 2

    let splitNumber (number: BigInteger) (at: int) = 
        let splitBase = 10I ** at
        let high = number / splitBase
        let low = number % splitBase
        (high, low)

    // +*+? is my Karatsuba symbol
    let rec (+*+?) (a: BigInteger) (b: BigInteger) = 
        let mutable max = 0I
        
        match a with
        | a when a > b  -> max <- a
        | _             -> max <- b

        match max with
        | max when max < 10I    -> a * b
        | _                     -> karatsubaRec max a b
    
    and karatsubaRec max a b = 
        let midPoint        = dividePoint max
        
        let (highA, lowA)   = splitNumber a midPoint
        let (highB, lowB)   = splitNumber b midPoint
        
        let highs           = highA +*+? highB
        let lows            = lowA +*+? lowB
        let mids            = (lowA + highA) +*+? (lowB + highB)

        let firstTerm       = highs * (10I ** (2 * midPoint))
        let midTerm         = (mids - highs - lows) * (10I ** midPoint)
        let lastTerm        = lows

        firstTerm + midTerm + lastTerm